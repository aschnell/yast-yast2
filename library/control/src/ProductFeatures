## Path:    System/Yast2
## Description: Product Features
## Type:    yesno
## Default: yes
#
# Enable Auto-Login
#
ENABLE_AUTOLOGIN="yes"

## Type:    yesno
## Default: no
#
# Enable EVMS config
#
EVMS_CONFIG="no"

## Type:    string(simple,expert)
## Default: expert
#
# UI mode
#
UI_MODE="expert"

## Type:    string(simple,expert)
## Default: expert
#
# Warn about Incomplete Translations
#
INCOMPLETE_TRANSLATION_TRESHOLD="95"
