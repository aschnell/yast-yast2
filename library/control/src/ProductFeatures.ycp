/**
 * File:	modules/ProductFetures.ycp
 * Package:	installation
 * Summary:	Product features
 * Authors:	Anas Nashif <nashif@suse.de>
 *              Jiri Srain <jsrain@suse.cz>
 *
 * $Id: ProductFeatures.ycp 21507 2005-02-10 16:01:58Z mvidner $
 */

{
module "ProductFeatures";
textdomain "base";

import "Misc";
import "Mode";
import "Stage";

/**
 * Map of all features
 * See defaults map below for sample contents
 */
map<string,map<string,any> > features = nil;

/**
 * Features to be saved in running system
 * one item per feature, consists of key in file, section and option
 */
list<list<string> > features_to_save = [
    ["UI_MODE", "globals", "ui_mode"],
    ["ENABLE_AUTOLOGIN", "globals", "enable_autologin"],
    ["EVMS_CONFIG", "partitioning", "evms_config"],
    ["INCOMPLETE_TRANSLATION_TRESHOLD",
	"globals", "incomplete_translation_treshold"]
];

/**
 * Default values for features
 * two-level map, section_name -> [ feature -> value ]
 */
map<string,map<string,any> > defaults = $[
    "globals" : $[
	"incomplete_translation_treshold" : "95",
	"ui_mode" : "expert",
	"enable_autologin" : true,
	"language" : "",
	"skip_language_dialog" : false,
	"keyboard" : "",
	"runlevel" : "",
	"timezone" : "",
	"fam_local_only" : "never",
	"enable_firewall" : true,
	"firewall_enable_ssh" : false,
	"additional_kernel_parameters" : "",
	"flags" : [],
	"run_you" : true,
	"relnotesurl" : "",
    ],
    "partitioning" : $[
	"evms_config" : "no",
	"use_flexible_partitioning" : false,
	"flexible_partitioning" : $[],
    ],
    "software" : $[
	"software_proposal" : "selection",
	"selection_type" : `auto,
	"delete_old_packages" : true,
	"only_update_installed" : false,
	"packages_transmogrify" : "",
	"base_selection" : "",
	"packages" : [],
	"kernel_packages" : [],
	"addon_selections" : [],
	"inform_about_suboptimal_distribution" : false,
    ],
    "network" : $[
	"force_static_ip" : false,
    ],
];

// function pre-declarations

global string GetStringFeature (string section, string feature);


/**
 * Initialize default values of features
 * @param force boolean drop all settings which were set before
 */
void InitFeatures (boolean force) {
    if (! (force || features == nil))
	return;
    features = defaults;
}

/**
 * Set a feature section
 * Default values will be used where value not defined
 * @param section name string the name of the section
 * @param section_map a map containing data of the section
 */
global void SetSection(string section_name, map<string,any> section_map) {
    InitFeatures (false);
    y2debug("Setting section: %1", section_name);
    section_map = (map<string,any>)
	union (defaults[section_name]:$[], section_map);
    features[section_name] = section_map;
}

/**
 * Get a complete section for evaluation
 */
global map<string,any> GetSection(string section_name) {
    InitFeatures (false);
    return features[section_name]:$[];
}

/**
 * Save product features
 */
global void Save() {
    InitFeatures (false);
    foreach (list<string> f, features_to_save, {
	path p = .product.features + f[0]:"";
	string value = GetStringFeature (f[1]:"", f[2]:"");
	SCR::Write (p, value);
    });
    SCR::Write (.product.features, nil );     // flush
}

/**
 * Restore product features in running system
 */
global void Restore() {
    InitFeatures (true);
    list<string> available = (list<string>)SCR::Dir (.product.features);
    list<list<string> > to_read = filter (list<string> f, features_to_save, {
	contains (available, f[0]:"");
    });
    foreach (list<string> f, to_read, {
	any val = SCR::Read (.product.features + f[0]:"");
	if (val != nil)
	{
	    if (! haskey (features, f[1]:""))
		features[f[1]:""] = $[];
	    features[f[1]:"", f[2]:""] = val;
	}
    });
}

/**
 * Initialize the features structure if needed
 * Either read from /etc/YaST2/ProductFeatures or set default values
 */
global void InitIfNeeded () {
    if (features != nil)
	return;
    if (Stage::normal () || Stage::firstboot())
	Restore ();
    else
	InitFeatures (false);
}

/**
 * Get value of a feature
 * @param section string section of the feature
 * @param features string feature name
 * @return any the feature value
 */
global any GetFeature (string section, string feature) {
    InitIfNeeded ();
    any ret = features[section, feature]:nil;
    if (ret == nil)
	ret = "";
    return ret;
}

/**
 * Get value of a feature
 * @param section string section of the feature
 * @param features string feature name
 * @return string the feature value
 */
global string GetStringFeature (string section, string feature) {
    any value = GetFeature (section, feature);
    if (value == nil)
	return nil;
    else if (is (value, string))
	return (string)value;
    else if (is (value, boolean))
	return (boolean)value ? "yes" : "no";
    else
	return sformat ("%1", value);
}

/**
 * Get value of a feature
 * @param section string section of the feature
 * @param features string feature name
 * @return boolean the feature value
 */
global boolean GetBooleanFeature (string section, string feature) {
    any value = GetFeature (section, feature);
    if (value == nil)
	return nil;
    else if (is (value, boolean))
	return (boolean)value;
    else if (is (value, string) && tolower ((string)value) == "yes")
	return true;
    else
	return false;

}

/**
 * Set value of a feature
 * @param section string section of the feature
 * @param features string feature name
 * @param value any the feature value
 */
global void SetFeature (string section, string feature, any value) {
    InitIfNeeded ();
    if (! haskey (features, section))
	features[section] = $[];
    features[section, feature] = value;
}

/**
 * Set value of a feature
 * @param section string section of the feature
 * @param features string feature name
 * @param value string the feature value
 */
global void SetStringFeature (string section, string feature, string value) {
    SetFeature (section, feature, value);
}

/**
 * Set value of a feature
 * @param section string section of the feature
 * @param features string feature name
 * @param value boolean the feature value
 */
global void SetBooleanFeature (string section, string feature, boolean value) {
    SetFeature (section, feature, value);
}


/* EOF */
}
