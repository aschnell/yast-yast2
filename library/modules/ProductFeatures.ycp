/**
 * File:	modules/ProductFetures.ycp
 * Package:	installation
 * Summary:	Product features
 * Authors:	Anas Nashif <nashif@suse.de>
 *
 * $Id$
 */

{
    module "ProductFeatures";
    textdomain "base";

    import "Misc";
    import "Mode";

    /*********************************************************************
      *   Those are variables from the content file, we can override them
      *   in the control file
     *********************************************************************/

    /**
     * url of release notes (downloaded during internet test)
     */
    global string relnotesurl	= "";

    /**
     *  Run YOU during the Internet connection test.
     */
    global boolean run_you	= true;

    /**
     * list of flags from content file
     */
    global list flags		= [];


    /*********************************************************************
      *   Packages and Selections
     *********************************************************************/

    /**
     *  Control how the base selection is chosen:
     *    `fixed - the base selection is fixed (see fixed_base_selection below)
     *    `auto  - a base selection is automatically choosen (depending on disk
     *             space)
     */
    global symbol base_selection_choice		= `auto;

    /**
     *  Name of the fixed base selection.
     */
    global string fixed_base_selection		= "";


    /**
     * List of additional packages to add
     */
    global list<string> packages = [];

    /**
     *  If non-empty it will be called with WFM::CallFunction
     *	from Packages::ComputeSystemPackageList. Gets a list of packages
     *  as the one and only argument and must return a list of packages
     *  again.
     */
    global string packages_transmogrify		= "";


    /***
      * On some products, this can be turned on to prevent old
      * RPMs from being deleted (#35235)
      */
    global boolean delete_old_packages          = true;


    /**
     *  Should the software proposal inform about the fact that the user wants
     *  to install e.g. i386 on amd64.
     */
    global boolean inform_about_suboptimal_distribution = false;



    // Only an update, NOT an upgrade
    global boolean onlyUpdateInstalled = false;


    // How is the software proposal handled? What is it based on?
    // Possible values are: desktop|selection
    // Default is "selection"

    global string software_proposal = "selection";

    /*********************************************************************
      *   Partitioning
     *********************************************************************/

    /**
     * Flexible partitioning
     */
    global boolean flexible_partitioning = false;

    /**
     * EVMS config in YaST2
     */
    global boolean evms_config = false;

    /**
     * Flexible partitioning map
     */
    global map FlexiblePartitioning = $[];




    /*********************************************************************
      *   Misc. Features
      *
     *********************************************************************/


    /**
     * autologin is enabled by default.
     */
    global boolean enable_autologin = true;

    /**
     * Runlevel
     */
    global string runlevel = "";

    /**
     * Language
     */
    global string language = "";

    /**
     * Skip first Language if user has selected a langauge using
     * syslinux or selected the language using linuxrc.
     */
    global boolean skip_language_dialog = false;

    /**
     * Keyboard
     */
    global string keyboard = "";

    /**
     * Timezone
     */
    global string timezone = "";

    /**
     * needed for desktop product to use shorter timeslices.
     */
    global boolean use_desktop_scheduler = true;

    /**
     * The I/O scheduler selection
     */
    global string io_scheduler = "as";

    /**
     * Value of FAM_LOCAL_ONLY in /etc/sysconfig/fam
     * One of "kde", "gnome", "always", "never" (default)
     * Adds according to the default window manager
     * No need to save in the system
     */
    global string fam_local_only = "never";

    /**
     * UI mode
     * possible values: "simple" or "expert"
     */
    global string ui_mode = "expert";



    /**
     * Save product features
     */
    global void Save()
    {
        SCR::Write(.product.features.USE_DESKTOP_SCHEDULER,
                (use_desktop_scheduler) ? "yes" : "no" );
        SCR::Write(.product.features.UI_MODE, ui_mode );
        SCR::Write(.product.features.IO_SCHEDULER,
		io_scheduler);
        SCR::Write(.product.features.ENABLE_AUTOLOGIN,
                (enable_autologin) ? "yes" : "no" );
        SCR::Write(.product.features.EVMS_CONFIG,
                (evms_config) ?  "yes" : "no" );
        SCR::Write (.product.features, nil );     // flush
        return;

    }


    /**
     * Restore product features in running system
     */
    global void Restore()
    {
        use_desktop_scheduler = Misc::SysconfigRead(
                .product.features.USE_DESKTOP_SCHEDULER, "yes" ) == "yes";
	io_scheduler = Misc::SysconfigRead(
		.product.features.IO_SCHEDULER, "as");
	ui_mode = Misc::SysconfigRead(
		.product.features.UI_MODE, "expert");
        enable_autologin = Misc::SysconfigRead(
                .product.features.ENABLE_AUTOLOGIN, "yes") == "yes";
        evms_config = Misc::SysconfigRead(
                .product.features.EVMS_CONFIG, "no") == "yes";
        return;
    }


    // -----------------------------------------------
    /**
     * Constructor
     */
    global define void ProductFeatures () ``{
        if (Mode::normal && !Mode::initial && !Mode::cont)
            Restore();
        return;
    }

/* EOF */
}
