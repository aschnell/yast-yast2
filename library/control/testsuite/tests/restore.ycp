{

include "testsuite.ycp";
import "ProductFeatures";

map READ = $[
        "product" : $[
            "features" : $[
                "USE_DESKTOP_SCHEDULER" : "0",
                "ENABLE_AUTOLOGIN" : "yes",
                "EVMS_CONFIG" : "0",
                "IO_SCHEDULER" : "cfg",
                "UI_MODE" : "expert",
                "INCOMPLETE_TRANSLATION_TRESHOLD" : "95",
            ]
        ],
];

TEST (``(ProductFeatures::GetStringFeature ("globals", "ui_mode")), [READ], 0);
TEST (``(ProductFeatures::GetStringFeature ("globals", "enable_autologin")), [READ], 0);
TEST (``(ProductFeatures::GetBooleanFeature ("globals", "enable_autologin")), [READ], 0);


}
