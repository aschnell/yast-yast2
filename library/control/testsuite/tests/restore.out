Dir	.product.features: ["ENABLE_AUTOLOGIN", "EVMS_CONFIG", "INCOMPLETE_TRANSLATION_TRESHOLD", "IO_SCHEDULER", "UI_MODE", "USE_DESKTOP_SCHEDULER"]
Read	.product.features."UI_MODE" "expert"
Read	.product.features."ENABLE_AUTOLOGIN" "yes"
Read	.product.features."EVMS_CONFIG" "0"
Read	.product.features."INCOMPLETE_TRANSLATION_TRESHOLD" "95"
Return	expert
Return	yes
Return	true
